### PR TITLE
[statviz] Authorization in back-end

### PR DESCRIPTION
**Pull request by pylipp**
UPDATE: the statistics queries can be reached on the `/graphql` endpoint (for an authenticated user), and on the `/public` endpoint (public; only for testing; not effective in demo/prod).

You can test the authorization by starting GraphQL explorer (on `/graphql`) with a user from organisation 2:
- volunteer@samos.org can view data of base 3 but not base 2 / 4
- if an accepted agreement exists that involves base 3 and base 1, they can view data of base 1
- a warehouse volunteer cannot access beneficiary data, e.g. `beneficiaryDemographics`

~~Most importantly this PR removes the `/public` endpoint.
The statviz-related queries can be reached under the `/graphql` endpoint which requires an authenticated user to execute the request.
When this PR is merged to master, statviz dashboard will not work until the FE is updated to send a user JWT.~~

### CODE REVIEW

**Comment by HaGuesto** [comment]
@HaGuesto priority

**Comment by HaGuesto** [comment]
@pylipp you wanted to change something first before I review, right?

**Comment by pylipp** [comment]
@HaGuesto implementation is ready for review now but I want to add more tests

File: back/boxtribute_server/business_logic/statistics/queries.py [hunk 1]
```diff
@@ -9,28 +10,85 @@
...
 
 query = QueryType()
+public_query = QueryType()
 
 
 @query.field("beneficiaryDemographics")
-def resolve_beneficiary_demographics(*_, base_ids=None):
-    return compute_beneficiary_demographics(base_ids)
+def resolve_beneficiary_demographics(*_, base_id):
+    authorize_cross_organisation_access("beneficiary", "tag_relation", base_id=base_id)
```
[the remark this answers is not in the document]
Comment (HaGuesto) [reply back/boxtribute_server/business_logic/statistics/queries.py:18]: ACTION ITEM: Please check with Roanna

File: back/boxtribute_server/authz.py [hunk 2]
```diff
@@ -173,6 +175,91 @@ def authorize_for_reading_box(box) -> None:
...
+
+    involved_base_ids = user_base_ids + [base_id]
+    details = (
+        TransferAgreementDetail.select()
+        .join(
+            TransferAgreement,
+            on=(
+                (TransferAgreementDetail.transfer_agreement == TransferAgreement.id)
+                & (TransferAgreement.state == TransferAgreementState.Accepted)
+                & (TransferAgreementDetail.source_base << involved_base_ids)
```
Comment (HaGuesto) [inline back/boxtribute_server/authz.py:243]: please split this up @pylipp

File: back/boxtribute_server/business_logic/statistics/queries.py [hunk 3]
```diff
@@ -9,28 +10,85 @@
...
+    return compute_beneficiary_demographics(base_id)
+
+
+@public_query.field("createdBoxes")
+def public_resolve_created_boxes(*_, base_id):
+    return compute_created_boxes(base_id)
+
+
+@public_query.field("topProductsCheckedOut")
+def public_resolve_top_products_checked_out(*_, base_id):
```
Comment (HaGuesto) [inline back/boxtribute_server/business_logic/statistics/queries.py:83]: please validate the input here in the queries @pylipp

**Review by HaGuesto — APPROVED** [review]
approved from my side. I trust you that you will do the changes we discussed today before you merge @pylipp

File: back/boxtribute_server/business_logic/statistics/queries.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pylipp) [reply back/boxtribute_server/business_logic/statistics/queries.py:26]: https://boxwise.slack.com/archives/C04QEFG4V35/p1700772851369679?thread_ts=1700061999.315009&cid=C04QEFG4V35

**Pull request merged**